lib/kafka/sasl_gssapi_authenticator.rb

```diff
@@ -0,0 +1,75 @@
+require 'gssapi'
+
+module Kafka
+  class SaslGssapiAuthenticator
+    GSSAPI_IDENT = "GSSAPI"
+    GSSAPI_CONFIDENTIALITY = false
+
+    def initialize(connection:, logger:, sasl_gssapi_principal:, sasl_gssapi_keytab:)
+      @connection = connection
+      @logger = logger
+      @principal = sasl_gssapi_principal
+      @keytab = sasl_gssapi_keytab
+
+      initialize_gssapi_context
+    end
+
+    def authenticate!
+      proceed_sasl_gssapi_negotiation
+    end
+
+    private
+
+    def proceed_sasl_gssapi_negotiation
+      @encoder = @connection.encoder
+      @decoder = @connection.decoder
+
+      response = @connection.send_request(Kafka::Protocol::SaslHandshakeRequest.new(GSSAPI_IDENT))
+      unless response.error_code == 0 && response.enabled_mechanisms.include?(GSSAPI_IDENT)
+        raise Kafka::Error, "#{GSSAPI_IDENT} is not supported."
+      end
+
+      # send gssapi token and receive token to verify
+      token_to_verify = send_sasl_token
+
+      # verify incoming token
+      unless @gssapi_ctx.init_context(token_to_verify)
+        raise Kafka::Error, "GSSAPI context verification failed."
+      end
+
+      # we can continue, so send OK
+      @encoder.write([0,2].pack('l>c'))
+
+      # read wrapped message and return it back with principal
+      handshake_messages
+
+      # no error, return @socket
+      @socket
+    end
+
+    def handshake_messages
+      msg = @decoder.read(@decoder.int32)
+      raise Kafka::Error, "GSSAPI negotiation failed." unless msg
+      # unwrap with integrity only
+      msg_unwrapped = @gssapi_ctx.unwrap_message(msg, GSSAPI_CONFIDENTIALITY)
+      msg_wrapped = @gssapi_ctx.wrap_message(msg_unwrapped + @principal, GSSAPI_CONFIDENTIALITY)
+      @encoder.write(msg_with_size(msg_wrapped))
+    end
+
+    def msg_with_size(msg)
+      [msg.size].pack('l>') + msg
```

Comment:
Isn't this the same as [`Encoder#write_bytes`](https://github.com/zendesk/ruby-kafka/blob/master/lib/kafka/protocol/encoder.rb#L90-L97)?